### ISSUE TITLE
Generate errors through Commander from client code

### ISSUE THREAD

**Issue by hildjj** [tj/commander.js#1632]
As mentioned on the [Implied Options](https://github.com/tj/commander.js/issues/1358#issuecomment-953268153) bug, I would like to perform inter-option checking after the main checks are done in `parse`.  In these extra checks, I would like to throw `InvalidArgumentError`, and have the same behavior as if I had done so in a coercion routine.  In particular, I would like the exception to be caught, and depending on what I've done with `exitOverride` and `configureOutput`, either re-raised or printed appropriately and `process.exit()` called.

I'm open to lots of different designs.  Here's a starting point for discussion:

```js
const argv = new Command()
  .option("-a", "First things first")
  .option("-b", "Second class")
  .check(opts => {
    if (opts.a && opts.b) {
      throw new InvalidArgumentError("-a and -b are mutually exclusive");
    }
    return { foo: opts.a || opts.b }; // Optional: replaces opts if truthy value returned
  })
  .parse()
  .opts();
```

The `check` routine gets called right before `parse` returns, perhaps inside the existing catch block for options.  If `check` returns undefined, the existing opts object (which might have been modified by the `check` method) is returned by `opts()`, otherwise whatever `check` returns will be used for `opts()`.

**Comment by shadowspawn** [comment]
Interesting question. Before Commander v4 there was not much extra that Commander did, but now as you noted there is `exitOverride` and `configureOutput`. And `showHelpAfterError`. And in the future...

The `InvalidArgumentError` exception is a special case though. Commander uses the context of the option coercion routine to know the option name and includes that in the error message for you. That particular handling can't be done when the exception is thrown from other parts of the code.

- https://github.com/tj/commander.js/blob/43f4743864e2f670db5eebcf88c92aa4612c54f1/lib/command.js#L571

The life-cycle hook for `"preAction"` is called after `parse` and before an action handler, so is a potential place for putting code for checking options. (i.e.  rather than adding a new `.check()` method.)

(I will think about convenient calling signatures to make use of the Commander built-in behaviours. Haven't got a discussion alternative for that yet.)

**Comment by shadowspawn** [comment]
Related prior art:
- http://yargs.js.org/docs/#api-reference-checkfn-globaltrue

**Comment by shadowspawn** [comment]
Experimenting with simple `.error("message")`.  In style of  `console.error()`, and argparse `.error()`, and oclif `.error()`.

```
program.error('-a and -b are mutually exclusive');
```

A potential downside with simple "error" is the name may clash with option name used in legacy programs still using `. storeOptionsAsProperties(true)`.

**Comment by hildjj** [comment]
Can there be an optional `code` parameter?  Would this need to be called from `action`, or would it just exit/throw depending on `exitOverride`?

**Comment by shadowspawn** [comment]
> Can there be an optional code parameter?

Yes. Is that something you would use straight away? For unit tests, or for try/catch in calling code?

> Would this need to be called from action, or would it just exit/throw depending on exitOverride?

I was thinking just exit/throw depending on exitOverride.

**Comment by hildjj** [comment]
Yes, I'm using the error code in unit tests.  I call exitOverride and configureOutput (passing in a function that writes to a small stream.Transform class to simulates stdio).  I check the error code is the one I expect, and that some chunk of the text is what I expect.

Duplicating or renaming the current behavior of _displayError would work very well for me, which means also having an optional exitCode, and perhaps more direct control over whether the help gets output.

**Comment by shadowspawn** [comment]
Experiment: https://github.com/shadowspawn/commander.js/commit/c223fae03cd0e722c590ed76a1eaa7b16bb8fcc0

**Comment by hildjj** [comment]
I like your experiment.  One possible improvement would be to give `code` and `exitCode` default values.

**Comment by hildjj** [comment]
Did this make it into v9.0.0?

**Comment by shadowspawn** [comment]
No, or at least not yet. This issue has not generated any likes or comments from other people, and I was not working on it.

I'll take another look.

**Comment by hildjj** [comment]
Thanks!  I'm doing ok using the undocumented internal function, and I think I've got a test that will catch when it changes so I can fix it up later.  It would of course be better if I didn't have to release like that. :)

**Comment by shadowspawn** [comment]
Opened a draft PR: #1675

**Comment by shadowspawn** [comment]
`.error()` is included in prerelease Commander v9.0.0-1.

**Comment by shadowspawn** [comment]
Commander v9 has been released.
